File: java/129.sum-root-to-leaf-numbers.java
/*
 * @lc app=leetcode id=129 lang=java
 *
 * [129] Sum Root to Leaf Numbers
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    int total = 0;

    public int sumNumbers(TreeNode root) {
        if (root != null) sumNumberAlongPaths(root, 0);
        return total;
    }

    private void sumNumberAlongPaths(TreeNode root, int sum) {
        sum = 10 * sum + root.val;
        if (root.left == null && root.right == null) total += sum;
        if (root.left != null) sumNumberAlongPaths(root.left, sum);
        if (root.right != null) sumNumberAlongPaths(root.right, sum);
    }
}

File: go/129.sum-root-to-leaf-numbers.go
/*
 * @lc app=leetcode id=129 lang=golang
 *
 * [129] Sum Root to Leaf Numbers
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
	if root == nil {
		return 0
	}
    return dfs(root, 0)
}

func dfs(root *TreeNode, acc int) int {
	acc = 10 * acc + root.Val
	if root.Left == nil && root.Right == nil {
		return acc
	}
	acc2 := 0
	if root.Left != nil {
		acc2 += dfs(root.Left, acc)
	}
	if root.Right != nil {
		acc2 += dfs(root.Right, acc)
	}
	return acc2
}
// @lc code=end

